server/src/main/java/org/elasticsearch/rest/RestController.java

```diff
@@ -344,6 +345,12 @@ private void tryAllHandlers(final RestRequest request, final RestChannel channel
                         BytesRestResponse.
                             createSimpleErrorResponse(channel, BAD_REQUEST, "multiple values for single-valued header [" + name + "]."));
                     return;
+                } else if (name.equals(Task.TRACE_PARENT)) {
+                    String traceparent = distinctHeaderValues.get(0);
+                    if (traceparent.length() >= 55) {
+                        threadContext.putTransient(Task.TRACE_ID, traceparent.substring(3, 35));
```

Comment:
An alternative to TraceIdConverter and setting the trace.is as a transient header is to set `org.apache.logging.log4j.ThreadContext.put("trace.id", traceId);`. This is how the Java agent handles log correlation. The ecs-logging-java lib will automatically add the trace id to the logs then.

Some caveats:
* At the end of the request `org.apache.logging.log4j.ThreadContext#clearAll` has to be called
* If some work is passed off to a thread pool, the context has to be propagated manually. Not sure if that's something ES does and whether `ThreadContext` is already propagated.

Reply:
`ThreadContext`  from `org.elasticsearch.threadpool.ThreadPool` exists to make sure headers are propagated not just between spawn threads but also between network hops automatically. 

I think given the current effort required no (additional) coordination to clear on log4j's ThreadContext and relies on the ES native propagation mechanics should be the way forward.

Even when Java Agent comes OOTB with Elasticsearch.

Reply:
For non-transient headers, such as `traceparent`, that's for sure the way to go. I just want to make sure we consider it for `trace.id`.